README.md

```diff
@@ -321,6 +321,129 @@ end
+
+One approach is to use standards-based Web Components. Because all Javascript functionality, internal markup, and styles are contained within the shadow root of the Web Component, it minimizes the risk of unexpected side-effects regarding other components.
+
+Here's an example of how that would work. You'd define three files per component: the `ViewComponent` Ruby object, the Rails view template, and the corresponding Javascript implementation of the Web Component.
+
+`app/components/comment_component.rb`
+```ruby
+class CommentComponent < ApplicationComponent
+  def initialize(comment:)
+    @comment = comment
+  end
+
+  def commenter
+    @comment.user
+  end
+
+  def commenter_name
+    commenter.name
+  end
+
+  def avatar
+    commenter.avatar_image_url
+  end
+
+  def formatted_body
+    simple_format(@comment.body)
+  end
+end
+```
+
+`app/components/comment_component.html.erb`
+```erb
+<my-comment comment-id="<%= comment.id %>">
+  <time slot="posted" datetime="<%= comment.created_at.iso8601 %>"><%= comment.created_at.strftime("%b %-d") %></time>
+
+  <div slot="avatar"><img src="<%= avatar %>" /></div>
+
+  <div slot="author"><%= commenter_name %></div>
+
+  <div slot="body"><%= formatted_body %></div>
+</my-comment>
+```
+
+`app/components/comment_component.js`
+```js
+class Comment extends HTMLElement {
+  styles() {
+    return `
+      :host {
+        display: block;
+      }
+      ::slotted(time) {
+        float: right;
+        font-size: 0.75em;
+      }
+      .commenter { … }
+      .body { … }
+    `
+  }
+
+  connectedCallback() {
+    this.shadowRoot.innerHTML = `
+      <style>
+        ${this.styles()}
+      </style>
+      <slot name="posted"></span>
+      <div class="commenter">
+        <slot name="avatar"></slot> <slot name="author"></slot>
+      </div>
+      <div class="body">
+        <slot name="body"></slot>
+      </div>
+    `
+  }
+}
+customElements.define('my-comment', Comment)
+```
+
+Another approach is to use a framework like Stimulus and create a 1:1 mapping between a Stimulus controller and a component. We'll leave the controller and HTML markup as an exercise for the reader, but in order to load in Stimulus controllers from the `app/components` tree, you'll have to amend the Stimulus boot code in `app/javascript/packs/application.js` as follows:
+
+```js
+const application = Application.start()
+const context = require.context("controllers", true, /.js$/)
+const context_components = require.context("../../components", true, /_controller.js$/)
+application.load(
+  definitionsFromContext(context).concat(
+    definitionsFromContext(context_components)
+  )
+)
+```
+
+This will allow you to create files such as `app/components/widget_controller.js` and as long as the controller identifier matches the `data-controller` attribute in the component's HTML template, you're all set.
```

Comment:
```suggestion
This will allow you to create files such as `app/components/widget_controller.js`, where the controller identifier matches the `data-controller` attribute in the component's HTML template.
```